client/site/main.js

```diff
@@ -75,3 +134,167 @@ function kill() {
+                Atomics.store(memoryViewu32, ptr++, i);
+            }
+            resource_map.delete(id);
+            queue.write_i32([8, id]);
+        } else {
+            console.error("Requested resource not in resource_map, id: " + id)
+        }
+    } else if (type === 1) {
+        if (x[1] === 0) {
+            window.addEventListener('input', input);
+        } else {
+            window.removeEventListener('input', input);
+        }
+    }
+}
+
+resource_map = new Map();
+function upload_resource(data) {
+    resource_map.set(data[0] & 0xff00, data)
+    queue.write_i32(7, data.length);
+}
+
+let canvas = createCanvas();
+if (canvas === undefined) throw Error('canvas creation failed');
+memory = generateMemory();
+let memoryView32 = new Int32Array(memory.buffer);
+let memoryViewU32 = new Uint32Array(memory.buffer);
+let memoryView8 = new Int8Array(memory.buffer);
+
+spawnModules(canvas, memory);
+
+function wakeUpAt(addr) {
+    Atomics.notify(memory.buffer, addr, +Infinity);
+}
+
+const START_TIME = Date.now();
+let last_time = START_TIME;
+const fps_sampling_count = 5;
+let fps_sampling_n = 0;
+async function wakeLogic() {
+    if (connected) {
+        let x = new UInt32Array(4);
+        x[0] = 128;
+        x[1] = Atomics.read(memoryView32, SYNC_PLAYER_STATE);
+        x[2] = Atomics.read(memoryView32, SYNC_PLAYER_STATE + 1);
+        x[3] = Atomics.read(memoryView32, SYNC_PLAYER_STATE + 2);
+        ws.post(x.buffer);
+    }
+
+    const t = Date.now() - START_TIME;
+    if (++fps_sampling_n >= fps_sampling_count) {
+        document.getElementById('lfps').textContent = Math.round(100000 / ((t - last_time) / fps_sampling_count)) / 100;
+        last_time = t;
+        fps_sampling_n = 0;
+    }
+    Atomics.store(memoryView32, SYNCHRONIZATION_MEMORY, t);
+    Atomics.store(memoryView32, SYNC_MOUSE, Math.floor(mousex));
+    Atomics.store(memoryView32, SYNC_MOUSE + 1, Math.floor(mousey));
+    Atomics.notify(memoryView32, SYNCHRONIZATION_MEMORY, +Infinity);
+}
+
+function setup_ws() {
+    ws.addEventListener('open',  () => {
+        add_text('ws connection to ' + WS_URL + 'established');
+        connected = true;
+    });
+    ws.addEventListener('error', event => {
+        add_text('ws error occured: "' + event + '"');
+        connected = false;
+    });
+    ws.addEventListener('close', event => {
+        add_text('ws is closed now: ' + event);
+        connected = false;
+    });
+    ws.addEventListener('message', event => {
+        let data = new UInt32Array(event.data.buffer);
+        let type = data[0] & 255;
+        if (type === 10) {
+            upload_resource(data);
+        } else if (type === 11) {
+            Atomics.store(memoryView32, SYNC_OTHER_STATE, data[1]);
+            Atomics.store(memoryView32, SYNC_OTHER_STATE + 1, data[2]);
+            Atomics.store(memoryView32, SYNC_OTHER_STATE + 2, data[3]);
+        }
+    });
+}
+
+String.prototype.hashCode = function() {
```

Comment:
I personally dislike that kind of monkey-patching, in my experience this mostly causes avoidable confusion. Is there any reason why we can't just define another function for this?